src/python/pants/backend/debian/target_types.py

```diff
@@ -13,11 +13,11 @@
 from pants.util.docutil import doc_url
 
 
-class DebianControlFile(Sources):
+class DebianSources(Sources):
     required = True
-    expected_num_files = 1
```

Comment:
What would the other files be other than the control file?

Reply:
I have rethought how we think of the sources for the Debian packages. I was a bit resistant to the idea of sources, but then it actually turned out to be great. 

Given this declaration:

```
debian_package(
    name='sample-debian-package',
    description='Useful tools installed as Debian package.',
    packages=[':files'],
    sources=['sample-debian-package/**/*'],
    symlinks={"myapp": "pex-target-name"},
    install_prefix="/opt/mycompany/tools",
    tags=["tooling"],
)
```

We can specify what directories we may need to be included. We need DEBIAN/control for the very least, but there can be multiple directories from various places - so why not let user pick what will be included into the package I thought? So instead of the `DebianControlFile` I think of `DebianSources`.

An example directory:

```
$ tree sample-debian-package      
sample-debian-package
├── DEBIAN
│   └── control
├── empty
└── opt
    └── company
        └── data.txt
```

Sources makes it possible for us to pick what we want to include.

Reply:
Hm those should probably be done via the `Dependencies` field and people using `files` targets etc. 

Generally, Pants works a lot better when only one target has "ownership" of a file. For example, dependency inference becomes ambiguous if >2 Python targets "own" the same Python file. Of course, dependency inference isn't relevant here. But, generally, it's better to stick to using `files`/`resources` targets for random files, and restrict `debian_package`'s `sources` to only include files that are specific to Debian like the control file.

Thoughts? I don't know what types of files get used with Debian packages, so I may be off base.

Reply:
However, the problem now is that I need somehow get to the "root" of the `sources=['sample-debian-package/**/*']` which is what `dpkg-deb` needs.

So a user could specify:

```
sources=['sample-debian-package/DEBIAN/control', 'sample-debian-package/opt/**/*']
```

and now I have to get the root directory name, in this case, `sample-debian-package`. Would. I need to write a little helper that would iteratively go to parent.parent... until it's root or you have a smarter way to do this? An assumption is that user will pick sources from a single directory, not different directories; that is, a single directory that has DEBIAN/control under it (among possibly other things).

Does this make any sense?

Reply:
We could also make it simpler -- most often user would really need a path to the directory containing the Debian package metadata. 

So `sources=['sample-debian-package']` -- i.e., Pants, can you please put this directory and everything underneath it into the Debian package? That will be enough, really. The only requirement is that underneath this directory there is `DEBIAN/control.` That's it.

```
$ tree sample-debian-package      
sample-debian-package
├── DEBIAN
│   └── control
├── empty
└── opt
    └── company
        └── data.txt
```

Reply:
> Thoughts? I don't know what types of files get used with Debian packages, so I may be off base.

It's quite common to put configuration files so the directory may look like this:

```
$ tree sample-debian-package      
sample-debian-package
├── DEBIAN
│   └── control
├── empty
└── etc
    └── configs
        └── config.rc
```

So if user provides `sources=['sample-debian-package']`, we'll take everything under that directory. That's why I wanted to go away from the control file only -- I need the whole directory.

Reply:
Okay if they are Debian specific files, that sounds sensible to me. @benjyw what do you think? We've never supported something like a directory for sources nor expected that the format is a particular way, like that the control file must be in a certain location.

Should we keep the `sources` field and have users do `sources=['debian/*']`? Or make a new field like `debian_config='debian/'`? If we do that custom field, there is a way we can still have `./pants package path/to/control` still work. Even though the target would not have traditional "ownership" of those files because it's not `Sources`, we can use `SecondaryOwnerMixin` to associate those files with the field.

Reply:
Thanks for giving it a thought, @Eric-Arellano! Just to provide more context:

Given this declaration:

```
debian_package(
    name='sample-debian-package',
    description='Useful tools installed as Debian package.',
    packages=[':files'],
    sources=['sample-debian-package-debian-sources/**/*'],
) 
```

User can do `./pants package //:sample-debian-package` (using the `name` from the declaration) to build a Debian package. The standard way is that there is a single directory that will be packaged as a Debian package and it is expected that this directory contains `DEBIAN/control` file. That's the absolute minimum. 

The user may decide to put some extra helper files into it, if desired. Here there is `/etc` directory inside the directory to be packaged:

```
$ tree sample-debian-package      
sample-debian-package
├── DEBIAN
│   └── control
└── etc
    └── configs
        └── config.rc
```

Therefore, the whole directory `sources=['sample-debian-package-debian-sources/**/*']` must be available upon packaging. I am happy with how things work now in this draft PR, with the only exception that sources ideally should be limited to the single directory (with all its content). People who are familiar wouldn't be passing arbitrary files from various locations in the repository because they keep all their resources in a single directory, dedicated to a particular Debian package, but it would be nice to enforce the constraints anyway.

Reply:
Sorry, I missed this. I think it's totally fine to do `sources=['sample-debian-package-debian-sources/**/*']` if the assumption is that all those sources need to be present as-is when you run the debian packaging command. That way Pants doesn't need to know anything about the structure, such as the existence of `DEBIAN/control` . So for example if in the future debian added some new way of providing the metadata, say `NEW_DEBIAN/new_control` , then things would still work. Not that they are likely to do that, but it illustrates that the less special-case knowledge Pants knows the better...

Reply:
Unrelated, but what is `packages=`? could you just use the transitive closure of `dependencies=` for this?

Reply:
I'm really sorry for the delay on this one, @alexey-tereshenkov-oxb! Okay, then yes that sounds reasonable to have `sources` allow >1 file and to not expect a certain file extension.

The help message should probably mention this is meant to include all Debian related files, like DEBIAN/control and opt/configs/config.rc.

We do expect there to be a file `DEBIAN/control`, though, right? If so, it'd help to implement the method `validate_resolved_files` on `DebianSources` - raise `InvalidFieldException`.

> Unrelated, but what is packages=? could you just use the transitive closure of dependencies= for this?

Yeah, your discussion around the behavior of depending on python_distribution was still pending. I recommended to Alexey that they use a dedicated `packages` field based on `relocated_files` for now to sidestep the conversation, and then we can revisit it / fix it if we decide we want to use `dependencies` in general. I continue to recommend that while you're changing the semantics of dependencies.